src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3989,184 @@ bool   os::win32::_is_windows_server         = false;
+    warning("os::malloc() failed to allocate %ld bytes for GetFileVersionInfo buffer", version_size);
+    return;
+  }
+
+  if (!GetFileVersionInfo(kernel32_path, 0, version_size, version_info)) {
+    os::free(version_info);
+    warning("GetFileVersionInfo() failed: GetLastError->%ld.", GetLastError());
+    return;
+  }
+
+  if (!VerQueryValue(version_info, TEXT("\\"), (LPVOID*)&file_info, &len)) {
+    os::free(version_info);
+    warning("VerQueryValue() failed. Cannot determine Windows version.");
+    return;
+  }
+
+  _major_version = HIWORD(file_info->dwProductVersionMS);
+  _minor_version = LOWORD(file_info->dwProductVersionMS);
+  _build_number  = HIWORD(file_info->dwProductVersionLS);
+  _build_minor   = LOWORD(file_info->dwProductVersionLS);
+
+  os::free(version_info);
+}
+
+bool os::win32::is_windows_11_or_greater() {
+  // Windows 11 starts at build 22000 (Version 21H2) as per
+  // https://learn.microsoft.com/en-us/windows/release-health/windows11-release-information
+  return (windows_major_version() >= 10 && windows_build_number() >= 22000 && !IsWindowsServer());
+}
+
+bool os::win32::is_windows_server_2022_or_greater() {
+  // Windows Server 2022 starts at build 20348.169 as per
+  // https://learn.microsoft.com/en-us/windows/release-health/release-information
+  return (windows_major_version() >= 10 && windows_build_number() >= 20348 && IsWindowsServer());
+}
+
+DWORD os::win32::active_processors_in_job_object() {
+  BOOL is_in_job_object = false;
+  if (!IsProcessInJob(GetCurrentProcess(), nullptr, &is_in_job_object)) {
+    warning("IsProcessInJob() failed: GetLastError->%ld.", GetLastError());
+    return 0;
+  }
+
+  if (!is_in_job_object) {
+    return 0;
+  }
+
+  DWORD processors = 0;
+
+  LPVOID job_object_information = nullptr;
+  DWORD job_object_information_length = 0;
+
+  if (!QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, nullptr, 0, &job_object_information_length)) {
+    DWORD last_error = GetLastError();
+    if (last_error == ERROR_INSUFFICIENT_BUFFER) {
+      DWORD group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
+
+      job_object_information = os::malloc(job_object_information_length, mtInternal);
+      if (job_object_information != nullptr) {
+          if (QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
+            DWORD groups_found = job_object_information_length / sizeof(GROUP_AFFINITY);
+            if (groups_found != group_count) {
+              warning("Unexpected processor group count->%ld. Expected %ld processor groups.", groups_found, group_count);
+              assert(false, "Unexpected group count");
+            }
+
+            GROUP_AFFINITY* group_affinity_data = ((GROUP_AFFINITY*)job_object_information);
+            for (DWORD i = 0; i < groups_found; i++) {
+              processors += population_count(group_affinity_data[i].Mask);
```

Comment:
you dont need the dereference, simply increment the pointer '... ;i++, group_affinity_data++) {' and then get to the mask via 'group_affinity_data->mask'

Reply:
The same pointer address is dereferenced using either style. I will leave it as is unless I get additional feedback upstream about this.

Reply:
It's not a style thing! The current code does more math (an additional multiply), you have spent more time checking that all the compilers will make the change for you than it would take to change the code.

nit: I strongly suggest to other C++ developers to follow the method where you increment the pointer

Reply:
> It's not a style thing! The current code does a multiply instead of an add to calculate the address likely using an additional register in the process, you have spent more time checking that all the compilers will make the change for you than it would take to change the code.
> 
> nit: I strongly suggest to other C++ developers to follow the method where you increment the pointer

Yes, I did spend some time investigating the generated code - simply changing the code without understanding the underlying point is not ideal in my view given my understanding that this was a standard array access pattern. I will change the array access style as you have recommended.